File: internal/handlers/file_handler.go
package handlers

import (
	"fmt"
	"net/http"
	"os"
	"path/filepath"
	"strings"

	"chat-backend/internal/config"
	"chat-backend/internal/database"

	"github.com/gin-gonic/gin"
	"github.com/google/uuid"
)

type FileHandler struct {
	db *database.Database
}

func NewFileHandler(db *database.Database) *FileHandler {
	return &FileHandler{db: db}
}

func (h *FileHandler) UploadFile(c *gin.Context) {
	cfg := config.Load()

	// Parse multipart form
	file, err := c.FormFile("file")
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "No file provided"})
		return
	}

	// Check file size
	if file.Size > cfg.MaxFileSize {
		c.JSON(http.StatusBadRequest, gin.H{"error": "File too large"})
		return
	}

	// Create upload directory if it doesn't exist
	if err := os.MkdirAll(cfg.UploadDir, os.ModePerm); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to create upload directory"})
		return
	}

	// Generate unique filename
	ext := filepath.Ext(file.Filename)
	filename := fmt.Sprintf("%s%s", uuid.New().String(), ext)
	filePath := filepath.Join(cfg.UploadDir, filename)

	// Save file
	if err := c.SaveUploadedFile(file, filePath); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to save file"})
		return
	}

	// URL that frontend uses as getBaseUrl() + file_url -> /api/v1/files/filename
	fileURL := fmt.Sprintf("/api/v1/files/%s", filename)

	c.JSON(http.StatusOK, gin.H{
		"file_url": fileURL,
		"filename": filename,
		"size":     file.Size,
		"type":     file.Header.Get("Content-Type"),
	})
}

func (h *FileHandler) ServeFile(c *gin.Context) {
	cfg := config.Load()
	filename := c.Param("filename")
	filePath := filepath.Join(cfg.UploadDir, filename)

	// Security: prevent directory traversal
	if !filepath.IsAbs(filePath) {
		filePath = filepath.Join(cfg.UploadDir, filepath.Base(filename))
	}

	file, err := os.Open(filePath)
	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "File not found"})
		return
	}
	defer file.Close()

	// Inline for images so they display in chat; attachment for others
	ext := strings.ToLower(filepath.Ext(filename))
	inlineExts := map[string]bool{".jpg": true, ".jpeg": true, ".png": true, ".gif": true, ".webp": true, ".svg": true}
	if inlineExts[ext] {
		c.Header("Content-Disposition", "inline")
	} else {
		c.Header("Content-Disposition", fmt.Sprintf("attachment; filename=%s", filename))
	}
	c.File(filePath)
}

File: src/handlers/fileHandler.js
'use strict';

const path = require('path');
const fs = require('fs');
const multer = require('multer');
const { loadConfig } = require('../config');

const config = loadConfig();
const uploadDir = config.uploadDir;

// Ensure upload directory exists
if (!fs.existsSync(uploadDir)) {
  fs.mkdirSync(uploadDir, { recursive: true });
}

// ─── Multer setup ─────────────────────────────────────────────────────────────

const storage = multer.diskStorage({
  destination: (_req, _file, cb) => {
    cb(null, uploadDir);
  },
  filename: (_req, file, cb) => {
    const ext = path.extname(file.originalname);
    const unique = `${Date.now()}-${Math.round(Math.random() * 1e9)}${ext}`;
    cb(null, unique);
  },
});

const upload = multer({
  storage,
  limits: { fileSize: config.maxFileSize || 10 * 1024 * 1024 }, // 10 MB
});

// ─── Handlers ─────────────────────────────────────────────────────────────────

async function uploadFile(req, res) {
  try {
    if (!req.file) {
      return res.status(400).json({ error: 'No file uploaded' });
    }

    const { filename, originalname, size, mimetype } = req.file;

    return res.json({
      url: `/api/v1/files/${filename}`,
      filename,
      original_name: originalname,
      size,
      content_type: mimetype,
    });
  } catch (err) {
    console.error('uploadFile error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

const IMAGE_EXTENSIONS = new Set(['.jpg', '.jpeg', '.png', '.gif', '.webp', '.svg']);

async function serveFile(req, res) {
  try {
    const filename = req.params.filename;

    // Prevent path traversal
    if (!filename || filename.includes('..') || filename.includes('/') || filename.includes('\\')) {
      return res.status(400).json({ error: 'Invalid filename' });
    }

    const filePath = path.join(uploadDir, filename);

    // Verify the resolved path is still within uploadDir
    const resolved = path.resolve(filePath);
    if (!resolved.startsWith(path.resolve(uploadDir))) {
      return res.status(400).json({ error: 'Invalid filename' });
    }

    if (!fs.existsSync(resolved)) {
      return res.status(404).json({ error: 'File not found' });
    }

    const ext = path.extname(filename).toLowerCase();
    if (IMAGE_EXTENSIONS.has(ext)) {
      res.setHeader('Content-Disposition', 'inline');
    } else {
      res.setHeader('Content-Disposition', `attachment; filename="${filename}"`);
    }

    return res.sendFile(resolved);
  } catch (err) {
    console.error('serveFile error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

module.exports = {
  upload,
  uploadFile,
  serveFile,
};
